Telegram charlen fix

sumwatt opened the pull request:
This is my first stab at Issue #1750 

The code change will chunk the text message into 4096 byte elements and loop over the elements and submit to Telegram.

cantino commented on app/models/agents/telegram_agent.rb:
```diff
@@ -79,7 +79,11 @@ def receive_event(event)
       params[:chat_id] = interpolated['chat_id']
       params[:parse_mode] = interpolated['parse_mode'] if interpolated['parse_mode'].present?
-      HTTMultiParty.post telegram_bot_uri(method), query: params
+      txt_message = params[:text].scan /.{1,4096}\W/m
```
What purpose does the `\W` serve here?

sumwatt replied:
It breaks the line on a non-word character. This may not be entirely necessary but was done more for continuity of the message.

knu replied:
Not to break in the middle of a word, I guess, but I suspect the last word would be lost if the text ends with a letter (`\w`).

sumwatt replied:
You're right - my initial test ended with a non-letter. Yours looks close. I did a quick test of 

/\G(?:\w{4096}|.{1,4096}(?<=\s)|.{1,4096}(?=\b|\z))/m


`2.3.0 :001 > blah = "this is a test nothing other than that"`
 `=> "this is a test nothing other than that" `
`2.3.0 :002 > blah.length`
` => 38 `
`2.3.0 :003 > blah.scan(/\G(?:\w{38}|.{1,38}(?<=\s)|.{1,38}(?=\b|\z))/m)`
` => ["this is a test nothing other than ", "that"] `

cantino commented:
Thanks @sumwatt. You'll also need to update `telegram_agent_spec.rb`. Let me know if you need a hand.

knu commented on app/models/agents/telegram_agent.rb:
```diff
@@ -79,7 +79,11 @@ def receive_event(event)
       params[:parse_mode] = interpolated['parse_mode'] if interpolated['parse_mode'].present?
-      HTTMultiParty.post telegram_bot_uri(method), query: params
+      txt_message = params[:text].scan /.{1,4096}\W/m
+      txt_message.each do | message |
```
You could do `params[:text].scan(pattern) do |message| ... end` instead of these two lines.

sumwatt replied:
I'll take care of that - I started going at this another direction so the extra variable assignment was a bit of an artifact.

sumwatt commented:
- Updated telegram_agent to remove the extraneous variable assignment. 
- Dumbed down the regex to do a hard split at 4096 characters until I can get the regex pattern working correctly
- Added additional test for splitting text to telegram_agent_spec.rb.... **Note** this is currently failing as I'm not quite sure how the stubbing process is working. It looks like it just takes whatever input it is given and assumes it is correct without actually invoking the underlying agent? I could possibly shoe-horn the text into the 'processes multiple events properly' test but it feels kind of kludgy....

dsander commented:
@sumwatt The tests are failing because currently `send_telegram_message` is [stubbed out completely](https://github.com/cantino/huginn/pull/1793/files#diff-74f5e9e92b397adf3ee35165e169d239L18) and the message splitting is happening in that method.
I think it would be a good idea to switch the stubs from `send_telegram_message` to `HTTMultiParty.post`, that would allow you to verify the message splitting and also ensure the rest of the method works as expected.
```
stub(HTTMultiParty).post do |url, args|
  @http_posts << { url => args[:query] }
end
```

iecompattest commented:
# [hi very good article
hi very good article HI very good a](

- 

#1. url

)
![ceresole-reale-1600x900-italy-mountains-lake-hills-clouds-6514](https://cloud.githubusercontent.com/assets/3653557/20697508/94085490-b5ae-11e6-94b7-4ea02905e7d1.jpg)
rticle

sumwatt commented on Gemfile.lock:
```diff
@@ -700,7 +700,7 @@ DEPENDENCIES
 
 RUBY VERSION
-   ruby 2.3.1p112
+   ruby 2.2.2p95
```
I will. having hard drive issues and it is getting ready to die. so I had to push before I lost it.

Maroo-b commented:
@sumwatt are you still working on this PR ?

dsander commented:
The feature was implemented in #2171
